File: Server/src/main/java/bgu/spl171/net/impl/TFTP/EncoderDecoders/StringEncoderDecoder.java
package bgu.spl171.net.impl.TFTP.EncoderDecoders;

import bgu.spl171.net.impl.TFTP.CommandEncoders.EncoderUtils;
import bgu.spl171.net.impl.TFTP.Commands.Command;
import bgu.spl171.net.impl.TFTP.Commands.ReadRequest;

import java.util.ArrayList;
import java.util.List;

public class StringEncoderDecoder {
    List<Byte> stringBuffer = new ArrayList<>();

    public String decodeNextByte(byte nextByte) {
        if (nextByte == 0) {
            String returnValue = getString();
            stringBuffer.clear();
            return returnValue;
        }
        return null;
    }

    private String getString() {
        byte[] bytes = new byte[stringBuffer.size()];
        for (int i = 0; i < bytes.length; i++) {
            bytes[i] = stringBuffer.get(i);
        }
        return new String(bytes);
    }
    public byte[] encode(String str){
        return EncoderUtils.mergeByteArray(str.getBytes(),new byte[]{0});
    }
}

File: java to c++ encdec/StringEncoderDecoder.cpp
#include "StringEncoderDecoder.h"

namespace bgu
{
	namespace spl171
	{
		namespace net
		{
			namespace impl
			{
				namespace TFTP
				{
					namespace EncoderDecoders
					{
						using bgu::spl171::net::impl::TFTP::CommandEncoders::EncoderUtils;
						using bgu::spl171::net::impl::TFTP::Commands::Command;
						using bgu::spl171::net::impl::TFTP::Commands::ReadRequest;

						std::wstring StringEncoderDecoder::decodeNextByte(char nextByte)
						{
							if (nextByte == 0)
							{
								std::wstring returnValue = getString();
								stringBuffer.clear();
								return returnValue;
							}
							return L"";
						}

						std::wstring StringEncoderDecoder::getString()
						{
							std::vector<char> bytes(stringBuffer.size());
							for (int i = 0; i < bytes.size(); i++)
							{
								bytes[i] = stringBuffer[i];
							}
							return std::wstring(bytes);
						}

						std::vector<char> StringEncoderDecoder::encode(const std::wstring &str)
						{
							return EncoderUtils::mergeByteArray(str.getBytes(),std::vector<char>{0});
						}
					}
				}
			}
		}
	}
}
